kale/embed/uncertainty_fitting.py

```diff
@@ -0,0 +1,121 @@
+from typing import List, Optional, Tuple
+
+import numpy as np
+import pandas as pd
+from yacs.config import CfgNode
+
+from kale.interpret.uncertainty_quantiles import quantile_binning_and_est_errors
+from kale.loaddata.tabular_access import load_csv_columns
+from kale.predict.uncertainty_binning import quantile_binning_predictions
+from kale.prepdata.tabular_transform import apply_confidence_inversion
+
+
+def fit_and_predict(
+    landmark: int,
+    uncertainty_error_pairs: List[List],
+    ue_pairs_val: str,
+    ue_pairs_test: str,
+    num_bins: int,
+    config: CfgNode,
+    groundtruth_test_errors: bool,
+    save_folder: Optional[str] = None,
+) -> Tuple[pd.DataFrame, pd.DataFrame, pd.DataFrame]:
+    """
+    Loads (validation, testing data) pairs of (uncertainty, error) pairs and for each fold: uses the validation
+    set to generate quantile thresholds, estimate error bounds and bin the test data accordingly. Saves
+    predicted bins and error bounds to a csv.
+
+    Args:
+        landmark (int): Which landmark to perform uncertainty estimation on.
+        uncertainty_error_pairs (list[list]): List of lists describing the different uncertainty combinations to test.
+        ue_pairs_val (str): Path to validation pairs (uncertainty, error) data.
+        ue_pairs_test (str): Path to test pairs (uncertainty, error) data.
+        num_bins (int): Number of bins for quantile binning.
+        config (CfgNode): Configuration object with hyperparameters and other settings.
+        groundtruth_test_errors (bool): Whether ground truth errors are available for test data.
+        save_folder (str, optional): Path to folder to save results to. If None, results are not saved.
+
+    Returns:
+        all_uncert_boundaries (pd.DataFrame): A DataFrame with uncertainty boundaries for each fold and uncertainty pairing.
+        error_bound_estimates (pd.DataFrame): A DataFrame with estimated error bounds for each fold and uncertainty pairing.
+        all_testing_results (pd.DataFrame): A DataFrame with predicted test bin values for each fold and uncertainty pairing.
+    """
+
+    logger = logging.getLogger("q_bin")
+    invert_confidences = config["DATASET"]["CONFIDENCE_INVERT"]
+    # dataset = config.DATASET.DATA
+    num_folds = config["DATASET"]["NUM_FOLDS"]
+    combine_middle_bins = config["PIPELINE"]["COMBINE_MIDDLE_BINS"]
+
+    # Save results across uncertainty pairings for each landmark.
+    all_testing_results = pd.DataFrame(load_csv_columns(ue_pairs_test, "Testing Fold", np.arange(num_folds)))
+    error_bound_estimates = pd.DataFrame({"fold": np.arange(num_folds)})
+    all_uncert_boundaries = pd.DataFrame({"fold": np.arange(num_folds)})
+
+    for idx, uncertainty_pairing in enumerate(uncertainty_error_pairs):
+        uncertainty_category = uncertainty_pairing[0]
+        invert_uncert_bool = [x[1] for x in invert_confidences if x[0] == uncertainty_category][0]
+        uncertainty_localisation_er = uncertainty_pairing[1]
```

Comment:
Not sure what 'er' stands for

Reply:
er = error. Are you requesting to change to "error" for better readability?